[integration] Implement Multi-VO File Catalog Metadata Handling

martynia opened the pull request:
BEGINRELEASENOTES
The idea of the update is to allow to store user metadata in a VO dependant way. For indexed
metadata a SQL table name, which normally holds a meta name as a suffix, is now extended to include w VO suffix. This way each VO writes to their own SQL table. 
The [DataManagementSystem/DB/FileCatalogComponents/DirectoryMetadata.py](url) and 
[DataManagementSystem/DB/FileCatalogComponents/FileMetadata.py](url) now contain a "hook" which, when overwritten in a derived class generates a required VO-specific tbale name suffix.
ENDRELEASENOTES

fstagni commented on DataManagementSystem/DB/FileCatalogComponents/DirectoryMetadata.py:
```diff
@@ -8,6 +8,7 @@
 import os
 from DIRAC import S_OK, S_ERROR
 from DIRAC.Core.Utilities.Time import queryTime
+import Utilities
```
Use full imports. Here, and later.

fstagni commented on DataManagementSystem/DB/FileCatalogComponents/DirectoryMetadata.py:
```diff
@@ -24,25 +25,52 @@ def setDatabase(self, database):
+    :return: meta name
+    """
+
+    return meta
```
What about returning `S_ERROR("method to be overwritten")` ?

martynia replied:
This method is a hook, and it works "as is" w/o overwriting it. I added an underscore to the method name to make it clear that it is an internal method.

atsareg replied:
Why do you need at all getMetaName and getMetaNameSuffix in the FC classes ? Just use methods from the Utilities module. The same for the File metadata, the utility functions are the same.

atsareg replied:
The functions can get argument specifying the naming method (no modifiction, by group, by VO, by ...). This can be defined in the catalog configuration. Then you do not need inherited classes to just implement this function

fstagni commented on DataManagementSystem/DB/FileCatalogComponents/DirectoryMetadata.py:
```diff
@@ -24,25 +25,52 @@ def setDatabase(self, database):
     if not result['OK']:
       return result
-    if pname in result['Value'].keys():
+    if fqPname in result['Value'].keys():
```
Please avoid the `.keys()` methods (not python 3 friendly)

martynia replied:
changed it to read: 
`if fqPname in result['Value']:`

I don't see why the original code (present all over the places in the same file would be not Python3 friendly. The `in` operator works file in Python 3 even thought `.keys()` does not return a list.

fstagni commented on DataManagementSystem/DB/FileCatalogComponents/DirectoryMetadata.py:
```diff
@@ -585,14 +620,17 @@ def findDirIDsByMetadata(self, queryDict, path, credDict):
         # given metadata, no need to check it further
         del finalMetaDict[meta]
-
+    print "finalMetaDict ?",finalMetaDict
```
No `print`s please.

martynia replied:
Removed.

fstagni commented on DataManagementSystem/DB/FileCatalogComponents/FileMetadata.py:
```diff
@@ -194,13 +205,47 @@ def __getFileID( self, path ):
 
+  @staticmethod
+  def fqMetaName(metaName, credDict):
+    return metaName+'_'+credDict['group']
```
For the correct format please see pep8 rules (that we adopt with the exclusion of the spacing). Also travis complains (https://travis-ci.org/DIRACGrid/DIRAC/jobs/540768734)

martynia replied:
The method is obsolete. Removed.

fstagni commented on DataManagementSystem/DB/FileCatalogComponents/MultiVODirectoryMetadata.py:
```diff
@@ -0,0 +1,39 @@
+
+__RCSID__ = "$Id$"
+
+from DirectoryMetadata import DirectoryMetadata
```
No relative imports.

martynia replied:
Fixed.

fstagni commented on Interfaces/API/Dirac.py:
```diff
@@ -1262,6 +1282,119 @@ def getLfnMetadata(self, lfns, printOutput=False):
 
     return repsResult
 
+  #### User Metadata API ######################################################
```
What are the advantages of adding these methods?

martynia replied:
I'm not insisting here. I just wanted to add all the client-side API call to already exported server-side calls, assuming that someone you had exported them had a reason to do so.

fstagni commented on tests/Integration/Resources/Catalog/Test_UserMetadata.py:
```diff
@@ -0,0 +1,127 @@
+"""
+Test of multi-VO user metadata handling. Assumes a running Dirac instance with the (master?) FileCatalog
```
Can you run this in Jenkins?

martynia replied:
Will eventually do, this is the plan.

fstagni commented on tests/Integration/Resources/Catalog/Test_UserMetadata.py:
```diff
@@ -0,0 +1,127 @@
+"""
+Test of multi-VO user metadata handling. Assumes a running Dirac instance with the (master?) FileCatalog
+"""
+
```
BTW the spacing of this file is 4 instead of 2

martynia replied:
Fixed.

chaen commented on Interfaces/API/Dirac.py:
```diff
@@ -65,6 +66,25 @@ def parseArguments(args):
     argList += arg.split(',')
   return argList
 
+def fileCatalogAPIWrapper(func):
```
that's more a debugging function for you, not really something to go to the API

martynia replied:
Well, this is a wrapper to hide away all the logging code which is identical for most of the methods. Otherwise a lot of code repetition would occur.

atsareg replied:
Please, do add FC related stuff to Dirac.py. FileCatalog() is itself the API.

martynia replied:
OK, but there was some FC related stuff there already. I've just added more ;-). What should go and what should stay in Dirac.py ?

atsareg replied:
The only pure FC method exposed through Dirac.py is getAllReplicas() which I think was a mistake to introduce. All the others do something on top of the FC queries. So, I do not think you should add anything at all to Dirac.py

fstagni replied:
Please drop them @martynia

martynia replied:
I understand you don't want to pollute Dirac.py with FC stuff, but it would be very useful to manipulate  user metadata programmatically (and there is server code for that). So is there any suitable place to move the client APi I have written ?

martynia replied:
And the tests would be useless since they use the client methods.

martynia commented:
Your inheritance proposal would work in some cases but it would not work everywhere. Look at the 
`__expandMetaDictionary(self, metaDict, credDict)` method. It uses a line:
`suffix = self._getMetaNameSuffix(credDict)`
to identify the suffix. This cannot be done by changing the derived method parameter. At least not at the first glance. Let me think about it.

atsareg commented on DataManagementSystem/Service/FileCatalogHandler.py:
```diff
@@ -523,16 +535,41 @@ def export_removeMetadata(self, pathMetadataDict):
+                             else key:value for key, value in metaDict['Value'].iteritems()}
+      if metaDict['MetadataOwner']:
+        metaDict['MetadataOwner']= {key.replace(suffix,'') if key.endswith(suffix)
+                             else key:value for key, value in metaDict['Value'].iteritems()}
```
Should it be "else key:value for key, value in metaDict['MetadataOwner'].iteritems()}" ?

martynia replied:
Good spot !

atsareg commented on DataManagementSystem/Service/FileCatalogHandler.py:
```diff
@@ -523,16 +535,41 @@ def export_removeMetadata(self, pathMetadataDict):
+
+      if metaDict['MetadataType']:
+        metaDict['MetadataType']= {key.replace(suffix,'') if key.endswith(suffix)
+                                    else key:value for key, value in metaDict['Value'].iteritems()}
```
idem

atsareg commented on Resources/Catalog/FileCatalogClient.py:
```diff
@@ -292,7 +292,7 @@ def getFileUserMetadata(self, path, timeout=120):
     directory = "/".join(path.split("/")[:-1])
     rpcClient = self._getRPC(timeout=timeout)
-    result = rpcClient.getFileUserMetadata(path)
+    result = rpcClient.findDirectoriesByMetadata(path)
```
Is it intentional ?

martynia replied:
Not sure why this happened. It clearly look like my commit.

atsareg commented:
You use user group as the metadata specification. While the intention is to use per VO metadata. Imagine the practical case: VO production manager produces data and users access it. The metadata will belong to vo_prod group and the user will look it up with the vo_user group and will not find any data.

atsareg commented on DataManagementSystem/DB/FileCatalogComponents/DirectoryMetadata.py:
```diff
@@ -248,7 +278,8 @@ def removeMetadata(self, dpath, metadata, credDict):
     """ Set an meta parameter - metadata which is not used in the the data
-        search operations
+        search operations.
+        Modified by JM: make metaName Dirac group aware.
```
Comments like that are suitable for release notes but for the in-code documentation

martynia replied:
OK

atsareg commented on DataManagementSystem/DB/FileCatalogComponents/DirectoryMetadata.py:
```diff
@@ -508,10 +539,14 @@ def __findSubdirMissingMeta(self, meta, pathSelection):
     """ Expand the dictionary with metadata query
     """
-    result = self.getMetadataFields(credDict)
+    result = self.getMetadataFields(credDict) # this now returns a fully qualified meta names, unlike the query !
```
Again this is a comment reflecting this particular release event, not relevant in the long term

atsareg commented on DataManagementSystem/DB/FileCatalogComponents/DirectoryMetadata.py:
```diff
@@ -585,14 +620,17 @@ def findDirIDsByMetadata(self, queryDict, path, credDict):
+      fqmeta = self.getMetaName(meta, credDict) # expand table name
+      # use fqmeta whet it is meant to be a table name
+      result = self.__checkDirsForMetadata(fqmeta, metaDict[meta], pathString)
+      print "__checkDirsForMetadata result", result
```
This print is also removed ?

atsareg commented on DataManagementSystem/DB/FileCatalogComponents/FileMetadata.py:
```diff
@@ -143,9 +151,12 @@ def setMetadata( self, path, metadict, credDict ):
 
-  def removeMetadata( self, path, metadata, credDict ):
-    """ Remove the specified metadata for the given file
+  def removeMetadata( self, path, metadata_r, credDict ):
```
metadata_r variable name is not following naming convention (rawMetadat ?)

martynia commented:
> You use user group as the metadata specification. While the intention is to use per VO metadata. Imagine the practical case: VO production manager produces data and users access it. The metadata will belong to vo_prod group and the user will look it up with the vo_user group and will not find any data.

Hi,

The code I'm working on now has a following statement, when resolving the suffix (we have discussed that during the meeting):

```python 
def _getMetaNameSuffix(self, credDict):
    """
    Get a VO specific suffix from user credentials.

    :param credDict: user credentials
    :return: VO specific suffix
    """
    vo = Registry.getGroupOption(credDict['group'], 'VO')
    return '_' + vo
```

atsareg commented on Resources/Catalog/FileCatalog.py:
```diff
@@ -293,6 +293,7 @@ def w_execute( self, *parms, **kws ):
     else:
       # FIXME: Return just master result here. This is temporary as more detailed
       # per catalog result needs multiple fixes in various client calls
+      # JM non-lfn methods would return {} if not a master...
```
Do not make personal comments please

fstagni commented:
Did you close this to re-open in v6r22?

atsareg commented:
No, this was closed by mistake

atsareg commented:
Though we have to decide which release it goes to. The changes are relatively contained to go into a patch. But they should be Jenkins-tested. Also, I do not see the need to make special FC components, everything can be in the original component code. This will not speed up testing or make the release safer because most of the modifications are in the original components anyway.

chaen commented:
Hint1: v7r1
Hint2: I advocate again that original components should be left untouched as much as possible

andresailer commented:
Maybe @martynia could push his updates?

martynia commented:
> Maybe @martynia could push his updates?

I would do it, need to address some comments still. I agree that we should discuss the current status, not the history.

atsareg commented on DataManagementSystem/DB/FileCatalogComponents/DirectoryMetadata.py:
```diff
@@ -57,26 +85,27 @@ def addMetadataField(self, pname, ptype, credDict):
 
-  def deleteMetadataField(self, pname, credDict):
-    """ Remove metadata field
+  def deleteMetadataField(self, r_pname, credDict):
```
r_pname -> rawPname

martynia commented:
I addressed most of the comments you made, less:

- `Dirac.py` API extension. This is still to be modified and moved elsewhere,

- moved the integration tests about, but still not in Jenkins (work in progress)

I defined both 'hook' methods to be private, keeping the inheritance tree, as before. One of them ('the suffix stripper') is called from the Handler -making it private required some more changes in the code (look for `strip_suffix` default parameter in some methods).

There might still be some minor things to change, but I would appreciate your comments about the concept first.

 Thanks, JM

martynia commented:
Hi,

I committed my changes last Friday and I was wondering if your have been notified. My other concern is that
Travis seemed to run the old code:

https://travis-ci.org/DIRACGrid/DIRAC/jobs/540768734 <https://travis-ci.org/DIRACGrid/DIRAC/jobs/540768734>

Could you give it (Travis) a kick ? Alternatively I could commit a small change to activate it.

   best, JM


> On 5 Jun 2019, at 14:10, Andrei Tsaregorodtsev <notifications@github.com> wrote:
> 
> @atsareg commented on this pull request.
> 
> In DataManagementSystem/DB/FileCatalogComponents/DirectoryMetadata.py <https://github.com/DIRACGrid/DIRAC/pull/4103#discussion_r290729141>:
> 
> >      if not result['OK']:
>        return result
>  
>      return S_OK("Added new metadata: %d" % metadataID)
>  
> -  def deleteMetadataField(self, pname, credDict):
> -    """ Remove metadata field
> +  def deleteMetadataField(self, r_pname, credDict):
> r_pname -> rawPname
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/DIRACGrid/DIRAC/pull/4103?email_source=notifications&email_token=AB2KE7J2P6E3XMNXJ2ZXVSTPY63NXA5CNFSM4HSINCOKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB2UZGSA#pullrequestreview-245994312>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AB2KE7MLD2AYAHX7F7Y4PYDPY63NXANCNFSM4HSINCOA>.
>

andresailer commented on DataManagementSystem/DB/FileCatalogComponents/DirectoryMetadata.py:
```diff
@@ -57,26 +85,27 @@ def addMetadataField(self, pname, ptype, credDict):
 
     req = "CREATE TABLE FC_Meta_%s ( DirID INTEGER NOT NULL, Value %s, PRIMARY KEY (DirID), INDEX (Value) )" \
-        % (pname, valueType)
+        % (fqPname, valueType)
```
I think there are VOs with hyphens in their name, so the TableName now needs to be surrounded by backticks, doesn't it?

martynia replied:
This is not the only place which suffers from this problem. FileMetadata.py has similar statements all over the place. The problem existed even before, if a user decided to name his metadata using hyphens. I can bactick all the occurrences if you find it  a right way to go, although a really robust solution would be to use sqlalchemy, like in other Dirac modules. This would be a big change, though.

andresailer replied:
At the moment hyphens in metadata names are not supported, which people can find a workaround. It might be harder to rename VOs for that reason.
Maybe instead of adding backticks everywhere one can replace hyphens in the getMetaNameSuffix method?

andresailer commented on DataManagementSystem/DB/FileCatalogComponents/FileMetadata.py:
```diff
@@ -507,102 +550,102 @@ def __findFilesByMetadata( self, metaDict, dirList, credDict ):
-    fileMetaDict = dict( item for item in metaDict.items() if item[0] in fileMetaKeys )
-
+    fileMetaDict = dict(item for item in metaDict.items()
+                        if item[0] + self._getMetaNameSuffix(credDict) in fileMetaKeys)
```
```suggestion
                        if self._getMetaName(item[0], credDict) in fileMetaKeys)
```

martynia replied:
Yes, much cleaner.

andresailer commented on DataManagementSystem/DB/FileCatalogComponents/MultiVODirectoryMetadata.py:
```diff
@@ -0,0 +1,39 @@
+  def __init__(self, database=None):
+    DirectoryMetadata.__init__(self, database=database)
+
+  def _getMetaName(self, meta, credDict):
```
Maybe this and getMetaNameSuffix can be turned into a Mixin class and added to this and MultiVOFileMetadata

martynia replied:
Yes, could do that. We would need 2 mixing classes: BaseMixin (pass through) and DerivedMixin one to maintain the suffix adding and removal. Base FileMetadata and DirectoryMetadata classes would both inherit from the BaseMixin and the Multi VO counterparts would inherit from the DerivedMixin. If it is what you meant I can implement this. I'll obviously find suitable names for the mixin classes.

andresailer replied:
Yes. Reduces code duplication

martynia commented on DataManagementSystem/DB/FileCatalogComponents/DirectoryMetadata.py:
```diff
@@ -101,7 +130,11 @@ def getMetadataFields(self, credDict):
+    # strip the VO suffix, if required
+    if strip_suffix:
+      suffix = self._getMetaNameSuffix(credDict)
+      metaDict = {key.replace(suffix, ''): value for key, value in metaDict.iteritems()
```
I changed the statement to read:
```python
metaDict = {key.rsplit(suffix, 1)[0] if key.endswith(suffix)
                  else key: value for key, value in metaDict.iteritems()}
```
It solves a problem when a user defined meta name ends with the _ and vo name (like __gridpp)_
Addressing your second comment: for file metadata the original dict only uses _tablename_metaname_ keys, all of them are extended by __voname_. For directories also other keys are present which have to be kept. I think it is much safer to use the 'keep other keys' approach, as above for all cases.

andresailer replied:
If you always use the same approach, then maybe turn this into a function.

martynia replied:
In the brand new mixin ....

andresailer replied:
:+1:

martynia commented on DataManagementSystem/DB/FileCatalogComponents/DirectoryMetadata.py:
```diff
@@ -512,6 +557,10 @@ def __expandMetaDictionary(self, metaDict, credDict):
       return result
     metaTypeDict = result['Value']
+    suffix = self._getMetaNameSuffix(credDict)
+    # remove the suffix - wrong, creates problem with selections down the line.
```
It must have been a comment I left for myself and forgot to remove it.

martynia commented on DataManagementSystem/DB/FileCatalogComponents/FileMetadata.py:
```diff
@@ -13,442 +13,486 @@
-  def deleteMetadataField( self, pname, credDict ):
-    """ Remove metadata field
+  def deleteMetadataField(self, pname, credDict):
+    """ Remove metadata field (only from user's own group)
```
Sure

martynia commented on DataManagementSystem/DB/FileCatalogComponents/FileMetadata.py:
```diff
@@ -13,442 +13,486 @@
-  def removeMetadata( self, path, metadata, credDict ):
-    """ Remove the specified metadata for the given file
+  def removeMetadata(self, path, rawMetadata, credDict):
+    """ Remove the specified metadata for the given file (for user's own group only)
```
ditto

andresailer commented:
Side note: At the moment the Transformation InputDataAgent uses the host credentials to find files by metadata, which would have the "hosts" group associated to it, which doesn't belong to any VO.
One would have to make some adjustments to make the multiVO metadata also work nicely with a multiVO InputDataAgent.

martynia commented:
Thanks,
We will have to take care of that.

   best, JM

> On 18 Jun 2019, at 15:09, Andre Sailer <notifications@github.com> wrote:
> 
> Side note: At the moment the Transformation InputDataAgent uses the host credentials to find files by metadata, which would have the "hosts" group associated to it, which doesn't belong to any VO.
> One would have to make some adjustments to make the multiVO metadata also work nicely with a multiVO InputDataAgent.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/DIRACGrid/DIRAC/pull/4103?email_source=notifications&email_token=AB2KE7NZINJP2OOW5XAY7SDP3DUDBA5CNFSM4HSINCOKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODX6X7VI#issuecomment-503152597>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AB2KE7KIBRGKMN2K6U6JEALP3DUDBANCNFSM4HSINCOA>.
>

fstagni commented on DataManagementSystem/DB/FileCatalogComponents/DirectoryMetadata.py:
```diff
@@ -350,6 +384,17 @@ def getDirectoryMetadata(self, path, credDict, inherited=True, owndata=True):
+                       else key: value for key, value in metaOwnerDict.iteritems()}
+
+      metaTypeDict = {key.replace(suffix, '') if key.endswith(suffix)
+                      else key: value for key, value in metaTypeDict.iteritems()}
```
What about
```
def stripSuffix(aDict, suffix):
  return {key.replace(suffix, '') if key.endswith(suffix)
              else key: value for key, value in aDict.iteritems()}

if strip_suffix:
  metaDict = stripSuffix(metaDict, suffix)
  metaOwnerDict = stripSuffix(metaOwnerDict, suffix)
  metaTypeDict = stripSuffix(metaTypeDict, suffix)

```

martynia replied:
Not there anymore. Added mixing classes as suggested by Andre.

fstagni commented on DataManagementSystem/DB/FileCatalogComponents/DirectoryMetadata.py:
```diff
@@ -302,7 +336,7 @@ def getDirectoryMetaParameters(self, dpath, credDict, inherited=True, owndata=Tr
     return S_OK(metaDict)
 
-  def getDirectoryMetadata(self, path, credDict, inherited=True, owndata=True):
+  def getDirectoryMetadata(self, path, credDict, inherited=True, owndata=True, strip_suffix=False):
```
normally we use `camelCase` instead of `pascal_case`

fstagni commented on DataManagementSystem/DB/FileCatalogComponents/FileMetadata.py:
```diff
@@ -13,442 +13,486 @@
     """ Set an meta parameter - metadata which is not used in the the data
-        search operations
+        search operations.
+        Modified by JM: make metaName Dirac group aware.
```
...it seems you committed something in your local repo, but it didn't yet reach here.

fstagni commented on DataManagementSystem/DB/FileCatalogComponents/FileMetadata.py:
```diff
@@ -13,442 +13,486 @@
+    This method is a 'hook' and is meant to be overwritten by derived classes.
+
+    :param meta:  meta name
+    :param credDict: client credentials
```
```
    :param str meta:  meta name
    :param dict credDict: client credentials
```
(also later)

martynia replied:
These have been addressed in the last push. (mix-in)

fstagni commented on tests/Integration/DataManagementSystem/Test_UserMetadata.py:
```diff
@@ -0,0 +1,128 @@
+if __name__ == '__main__':
+  suite = unittest.defaultTestLoader.loadTestsFromTestCase(TestUserMetadataTestCase)
+  suite.addTest(unittest.defaultTestLoader.loadTestsFromTestCase(testMetadata))
+  testResult = unittest.TextTestRunner(verbosity=2).run(suite)
```
Add at the end here:
```
  sys.exit(not testResult.wasSuccessful())
```

martynia replied:
Done

fstagni commented:
Please rebase on `integration` branch.

martynia commented:
> Please rebase on `integration` branch.
Done. Waiting for tests to complete.

martynia commented:
Hi,

One topic which was (I hope) covered by my colleagues  during the last Dirac meeting is relocation of the File Catalog  user metadata API. 

Is the  directory:

Resources/Catalog/ 

a suitable place to move the user metadata stuff from Dirac.py ?

best, JM  

> On 4 Jul 2019, at 11:18, fstagni <notifications@github.com> wrote:
> 
> Please rebase on integration branch.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/DIRACGrid/DIRAC/pull/4103?email_source=notifications&email_token=AB2KE7JHRTOEPODO2IIRKALP5XFA3A5CNFSM4HSINCOKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZHAFUI#issuecomment-508429009>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AB2KE7JNIU2RG4XGNLVLZKDP5XFA3ANCNFSM4HSINCOA>.
>

chaen commented:
Hi,

yes they mentioned something about it, but I have not looked at it in detail yet.
I am not really sure what you are trying to achieve here. Can you elaborate please ?

chaen commented:
Maybe a bit more meat: `FileCatalog` is the API when you need to do things from within DIRAC.
If you want users to be exposed to some function, then `Dirac.py` makes sense. At the same time, it should not be too messy. So since there is already a `Job.py`, maybe we can create another file. 
But again, it all depends what you are trying to do

martynia commented:
Well,

I originally put client calls related to user metadata handling into Dirac.py. The client calls use existing RPC server-side calls, so I was not very innovative here.

The client operations serve both testing purposes (‘within Dirac', in your speak) and should be exposed to the users (or developers creating tools like data movers etc.), so they can manipulate data programmatically. If Dirac.py is too long (and IMHO it is), we could create another file in the same directory, i.e. UserMetadata.py or move it to the FileCatalog API. 

 If we drop the API completely that the only way users could manipulate metadata would be by using the File Catalog CLI. Tool developers would be stuck.

From the discussion on GitHub the opposition to keeping the API in Dirac.py is strong. So where move it to ?

Best, JM

> On 11 Jul 2019, at 14:38, chaen <notifications@github.com> wrote:
> 
> Maybe a bit more meat: FileCatalog is the API when you need to do things from within DIRAC.
> If you want users to be exposed to some function, then Dirac.py makes sense. At the same time, it should not be too messy. So since there is already a Job.py, maybe we can create another file.
> But again, it all depends what you are trying to do
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/DIRACGrid/DIRAC/pull/4103?email_source=notifications&email_token=AB2KE7IGO5PONWZNGHFVQZDP64ZUDA5CNFSM4HSINCOKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZWXC3Q#issuecomment-510488942>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AB2KE7KAHR5RILJTEA6EFWLP64ZUDANCNFSM4HSINCOA>.
>

chaen commented:
For developers the interface to the file catalog *is* `FileCatalog.py`. 
For users, I am not sure, our users do not use the metadata of DFC

martynia commented:
Removed FC API from Dirac.py + some bug fixes + autopep8

fstagni commented on DataManagementSystem/DB/FileCatalogComponents/MultiVOFileMetadata.py:
```diff
@@ -0,0 +1,17 @@
+
+  def __init__(self, database=None):
+    FileMetadata.__init__(self, database=database)
+    MultiVOMetaNameMixIn.__init__(self)
```
These inheritances are partly broken. You are in this situation right now:

```
In [1]: class MetaNameMixInNoObject():
   ...:   def __init__(self):
   ...:     print("I am MetaNameMixInNoObject")
   ...: 
   ...: 
   ...: class MultiVOMetaNameMixInNoObject(MetaNameMixInNoObject):
   ...:   def __init__(self):
   ...:     print("I am MultiVOMetaNameMixInNoObject")
   ...: 
   ...: 
   ...: class DirectoryMetadataNoObject(MetaNameMixInNoObject):
   ...:  def __init__(self):
   ...:   print("I am DirectoryMetadataNoObject")
   ...: 
   ...: 
   ...: class MultiVODirectoryMetadataNoObject(MultiVOMetaNameMixInNoObject, DirectoryMetadataNoObject):
   ...:   def __init__(self):
   ...:     print("I am MultiVODirectoryMetadataNoObject")
   ...:     DirectoryMetadataNoObject.__init__(self)
   ...:     MultiVOMetaNameMixInNoObject.__init__(self)
   ...: 
   ...: 
   ...: MultiVODirectoryMetadataNoObject()
   ...: 
I am MultiVODirectoryMetadataNoObject
I am DirectoryMetadataNoObject
I am MultiVOMetaNameMixInNoObject
```
As you see what's in `MetaNameMixInNoObject` is not executed. This is because you are using old style classes: http://python-history.blogspot.com/2010/06/method-resolution-order.html

This might not be a problem with the current code because there's nothing really done in  `MetaNameMixInNoObject`, but we shouldn't rely on that.


What should instead be done is the following (inheriting from `object` and using `super`:
```
In [1]: class MetaNameMixIn(object):
   ...:   def __init__(self):
   ...:     print("I am MetaNameMixIn")
   ...: 
   ...: 
   ...: class MultiVOMetaNameMixIn(MetaNameMixIn):
   ...:   def __init__(self):
   ...:     super(MultiVOMetaNameMixIn, self).__init__()
   ...:     print("I am MultiVOMetaNameMix")
   ...: 
   ...: 
   ...: class DirectoryMetadata(MetaNameMixIn):
   ...:   def __init__(self):
   ...:     super(DirectoryMetadata, self).__init__()
   ...:     print("I am DirectoryMetadata")
   ...: 

In [2]: class MultiVODirectoryMetadataWithSuper(MultiVOMetaNameMixIn, DirectoryMetadata):
   ...:   def __init__(self):
   ...:     super(MultiVODirectoryMetadataWithSuper, self).__init__()
   ...:     print("I am MultiVODirectoryMetadataWithSuper")
   ...: 

In [3]: MultiVODirectoryMetadataWithSuper()
I am MetaNameMixIn
I am DirectoryMetadata
I am MultiVOMetaNameMix
I am MultiVODirectoryMetadataWithSuper
```

Also, python 3 by default would use the MRO of new-style classes.

martynia replied:
Sorry for not replying straight away, but I was on holidays and I'm taking some days off next week again ...

I totally agree I should be using new style classes, I'll fix this.  However your example above is not quite accurate since I explicitly call base class inits:
https://github.com/DIRACGrid/DIRAC/blob/0d83eddd103d050edcec6abaf58b2ea32f3ba532/DataManagementSystem/DB/FileCatalogComponents/DirectoryMetadata.py#L16-L19

and also:

https://github.com/DIRACGrid/DIRAC/blob/0d83eddd103d050edcec6abaf58b2ea32f3ba532/DataManagementSystem/DB/FileCatalogComponents/MultiVOMetaNameMixIn.py#L15-L16

Another problem with `super` is that all `init` prototypes have to agree, and in our example we have an additional `database=`  argument in one of them. I could pass it as a default to another one but it would not look nice.

How do you see this ?

   best, JM

fstagni replied:
You can define `def __init__(self, *args, **kwargs):` for absorbing all the arguments.

martynia replied:
Hi, the inheritance tree has been updated.

fstagni commented:
@martynia please just fix https://travis-ci.org/DIRACGrid/DIRAC/jobs/572821909#L2693 in my view this would be the last change to be done here.

At the same time I am not the most expert in the code written here, so I'd like @atsareg @chaen and @andresailer to give a final review.

andresailer commented on DataManagementSystem/DB/FileCatalogComponents/DirectoryMetadata.py:
```diff
@@ -350,6 +362,13 @@ def getDirectoryMetadata(self, path, credDict, inherited=True, owndata=True):
         metaOwnerDict[meta] = 'OwnParameter'
 
+    if strip_suffix:
+      #      metaDict = {key.replace(suffix, '') if key.endswith(suffix)
```
It is not necessary to keep these commented lines

andresailer commented on DataManagementSystem/DB/FileCatalogComponents/DirectoryMetadata.py:
```diff
@@ -512,6 +531,9 @@ def __expandMetaDictionary(self, metaDict, credDict):
     if not result['OK']:
       return result
     metaTypeDict = result['Value']
+    # strip the suffix
```
That  comment is not adding anything useful

andresailer commented on DataManagementSystem/DB/FileCatalogComponents/DirectoryMetadata.py:
```diff
@@ -302,7 +314,7 @@ def getDirectoryMetaParameters(self, dpath, credDict, inherited=True, owndata=Tr
     return S_OK(metaDict)
 
-  def getDirectoryMetadata(self, path, credDict, inherited=True, owndata=True):
+  def getDirectoryMetadata(self, path, credDict, inherited=True, owndata=True, strip_suffix=False):
```
It would be helpful to have a description under which circumestances strip_suffix true or false should be used

martynia replied:
```strip_suffix=true```  is to be used when a call is meant to be in a call chain which returns data to a user, otherwise it should be ``` false``` to maintain backward compatibility since that method is used internally as well.

andresailer commented on DataManagementSystem/DB/FileCatalogComponents/DirectoryMetadata.py:
```diff
@@ -604,9 +626,9 @@ def findDirIDsByMetadata(self, queryDict, path, credDict):
       for meta, value in finalMetaDict.items():
         if value == "Missing":
-          result = self.__findSubdirMissingMeta(meta, pathSelection)
+          result = self.__findSubdirMissingMeta(fqmeta, pathSelection)
```
Should this still be `meta`, given that this comes from the loop variable two lines above?

martynia replied:
I think we need a line:
```python
fqmeta = self.getMetaName(meta, credDict)
```
as a first line in both loops, so ```python __findSubdirByMeta ``` looks into a correct table.

andresailer commented on DataManagementSystem/DB/FileCatalogComponents/DirectoryMetadata.py:
```diff
@@ -604,9 +626,9 @@ def findDirIDsByMetadata(self, queryDict, path, credDict):
+          result = self.__findSubdirMissingMeta(fqmeta, pathSelection)
         else:
-          result = self.__findSubdirByMeta(meta, value, pathSelection)
+          result = self.__findSubdirByMeta(fqmeta, value, pathSelection)
```
Should this still be meta, given that this comes from the loop variable four lines above?

martynia replied:
as above.

andresailer commented on DataManagementSystem/DB/FileCatalogComponents/MultiVOMetaNameMixIn.py:
```diff
@@ -0,0 +1,51 @@
+  def getMetaName(self, meta, credDict):
+    """
+    Return a fully-qualified metadata name based on client-suplied metadata name and
+    client credentials. User group is added to the metadata passed in.
```
It is the users VO that is added to the metadata name

martynia replied:
Yes.

andresailer commented on DataManagementSystem/DB/FileCatalogComponents/MultiVOMetaNameMixIn.py:
```diff
@@ -0,0 +1,51 @@
+  def stripSuffix(self, metaDict, credDict):
+    """
+    Strip the suffix from all keys which contain it, leaving other keys intact.
+    :param metaDict: original dict
```
Need a newline before the first `:param ...`

martynia replied:
done

andresailer commented on DataManagementSystem/DB/FileCatalogComponents/MultiVOMetaNameMixIn.py:
```diff
@@ -0,0 +1,51 @@
+    """
+
+    suffix = self.getMetaNameSuffix(credDict)
+    smetaDict = {key.rsplit(suffix, 1)[0]: value for key, value in metaDict.iteritems()
```
This removes all entries that do not end with suffix, which doesn't sound like what is intended given the docstring

martynia replied:
A uses only gets its own VO data, so stripping all the other keys is intended. The comment is wrong.

chaen commented:
I let @andresailer decide since it touches only File/Directory Metadata managers, which we do not use

martynia commented:
I addressed the comments a few days ago. Is it anything else which should be done on this ?
Thanks, JM

martynia commented on DataManagementSystem/DB/FileCatalogComponents/DirectoryMetadata/DirectoryMetadata.py:
```diff
@@ -27,22 +29,25 @@ def setDatabase(self, database):
-    if pname in result['Value'].keys():
-      if ptype.lower() == result['Value'][pname].lower():
+
+    if fqPname in result['Value'].keys():
```
OK

martynia commented on DataManagementSystem/DB/FileCatalogComponents/FileMetadata/FileMetadata.py:
```diff
@@ -37,56 +39,58 @@ def addMetadataField(self, pname, ptype, credDict):
       return result
-    if pname in result['Value'].keys():
-      if ptype.lower() == result['Value'][pname].lower():
+    if fqPname in result['Value'].keys():
```
OK

martynia commented on tests/Integration/DataManagementSystem/Test_UserMetadata.py:
```diff
@@ -0,0 +1,140 @@
+"""
+Test of multi-VO user metadata handling. Assumes a running Dirac instance with the (master?) FileCatalog
+"""
+
```
OK, added to other new files.

martynia commented on DataManagementSystem/DB/FileCatalogComponents/MultiVOMetaNameMixIn.py:
```diff
@@ -0,0 +1,52 @@
+""" DIRAC Multi VO MixIn class to manage file metadata and directory for multiple VO.
+"""
+
```
Done.

martynia commented on tests/Integration/DataManagementSystem/Test_UserMetadata.py:
```diff
@@ -0,0 +1,140 @@
+    # meta show
+    result = self.fc.getMetadataFields()
+    self.assertTrue(result['OK'])
+    self.assertDictContainsSubset({'JMMetaInt6': 'INT', }, result['Value']['FileMetaFields'])
```
It must have been a typo.

martynia commented on tests/Integration/DataManagementSystem/Test_UserMetadata.py:
```diff
@@ -0,0 +1,140 @@
+    # API call only
+    result = self.fc.getFileUserMetadata(self.lfn5)
+    self.assertTrue(result['OK'])
+    # 'Value': {'JMMetaInt6': 13L, 'JMMetaInt5': 12L, 'JMTestDirectory': 124L}
```
No, removed.

martynia commented on DataManagementSystem/DB/FileCatalogComponents/FileMetadata/FileMetadata.py:
```diff
@@ -37,56 +39,58 @@ def addMetadataField(self, pname, ptype, credDict):
-      return S_ERROR('The metadata %s is already defined for Directories' % pname)
+    # existing pnames are fully qualified, so
+    fqPname = self.getMetaName(pname, credDict)
+    if fqPname in result['Value'].keys():
```
Done.

martynia commented on tests/Integration/DataManagementSystem/Test_UserMetadata.py:
```diff
@@ -0,0 +1,140 @@
+  def test_AddQueryRemove(self):
+    result = self.dirac.getLfnMetadata(self.lfn5)
+    self.assertTrue(result['OK'])
+    self.assertTrue(self.lfn5 in result['Value']['Successful'].keys())
```
Done.

martynia commented on tests/Integration/DataManagementSystem/Test_UserMetadata.py:
```diff
@@ -0,0 +1,140 @@
+def random_dd(outfile, size_mb):
+  import os
+  with open(outfile, 'w') as f:
+    for i in range((int(size_mb) * 2 ** 20) / 512):
```
Fixed.

martynia commented on DataManagementSystem/DB/FileCatalogComponents/MetaNameMixIn.py:
```diff
@@ -0,0 +1,43 @@
+
+class MetaNameMixIn(object):
+
+  def __init__(self, *args, **kwargs):
```
Done.

fstagni commented:
Just rebase or merge the `integration` branch and then I think it can be merged.

martynia commented:
I have synced  my branch with integration right now. It says it could be merged after review.
Best, JM

andresailer commented on DataManagementSystem/Service/FileCatalogHandler.py:
```diff
@@ -524,16 +526,26 @@ def export_removeMetadata(self, pathMetadataDict):
+        Client gets metadata w/o the VO suffix.
     """
-    return gFileCatalogDB.dmeta.getDirectoryMetadata(path, self.getRemoteCredentials())
+    metaDict = gFileCatalogDB.dmeta.getDirectoryMetadata(path, self.getRemoteCredentials(), enableStripping=False)
```
This line (and corresponding debug output) has to be dropped, there is no point querying the database twice only for debug reasons

martynia commented:
@zmathe could you please review this pull request ? Thanks, JM

zmathe commented on DataManagementSystem/DB/FileCatalogComponents/DirectoryMetadata/DirectoryMetadata.py:
```diff
@@ -90,7 +96,7 @@ def deleteMetadataField(self, pname, credDict):
     return result
 
-  def getMetadataFields(self, credDict):
+  def getMetadataFields(self, credDict, enableStripping=False):
```
In order to keep backward compatibility I would set enableStripping=True
And probably better to use stripVO ?

martynia replied:
This particular method is used in the handler in which case the VO suffix has to be stripped of and multiple times internally in its class where the suffix has to be present. Flipping the default would require more typing but it would not change anything. The backward compatibility with the single VO system is not maintained by this default but by loading a correct plugin - _DirectoryMetadata_ or _MultiVODirectoryMetadata_ class. In a single VO case the suffix is "stripped" as well, but it is defined to be an empty string so stripping has no effect.

fstagni replied:
I think it's fine to keep it `False` as default there.

zmathe commented on DataManagementSystem/DB/FileCatalogComponents/DirectoryMetadata/DirectoryMetadata.py:
```diff
@@ -28,22 +30,25 @@ def setDatabase(self, database):
   def addMetadataField(self, pname, ptype, credDict):
     """ Add a new metadata parameter to the Metadata Database.
         pname - parameter name, ptype - parameter type in the MySQL notation
+        Modified to use fully qualified metadata names.
```
probably better to update the doc string. Please use
:param str pname: ....

martynia replied:
I updated the docstrigs in the methods I modified to my best knowledge. Since I haven't written the code in the first place, some params description might be incomplete or even incorrect.

zmathe commented on DataManagementSystem/DB/FileCatalogComponents/DirectoryMetadata/DirectoryMetadata.py:
```diff
@@ -211,7 +221,7 @@ def setMetadata(self, dpath, metadict, credDict):
 
   def removeMetadata(self, dpath, metadata, credDict):
-    """ Remove the specified metadata for the given directory
+    """ Remove the specified metadata for the given directory for users own VO
```
please add the method variables to the doc string

zmathe commented on DataManagementSystem/DB/FileCatalogComponents/DirectoryMetadata/DirectoryMetadata.py:
```diff
@@ -249,7 +261,7 @@ def removeMetadata(self, dpath, metadata, credDict):
   def setMetaParameter(self, dpath, metaName, metaValue, credDict):
     """ Set an meta parameter - metadata which is not used in the the data
-        search operations
+        search operations.
```
here as well

zmathe commented on DataManagementSystem/DB/FileCatalogComponents/DirectoryMetadata/DirectoryMetadata.py:
```diff
@@ -303,10 +315,11 @@ def getDirectoryMetaParameters(self, dpath, credDict, inherited=True, owndata=Tr
     return S_OK(metaDict)
 
-  def getDirectoryMetadata(self, path, credDict, inherited=True, owndata=True):
+  def getDirectoryMetadata(self, path, credDict, inherited=True, owndata=True, enableStripping=False):
```
I have the same comment regarding to the enableStripping variable as above

zmathe commented on DataManagementSystem/DB/FileCatalogComponents/FileMetadata/FileMetadata.py:
```diff
@@ -11,25 +11,27 @@
   def addMetadataField(self, pname, ptype, credDict):
     """ Add a new metadata parameter to the Metadata Database.
-        pname - parameter name, ptype - parameter type in the MySQL notation
+        pname - parameter name, ptype - parameter type in the MySQL notation.
```
please use the proper doc string for the method variable

zmathe commented on DataManagementSystem/DB/FileCatalogComponents/FileMetadata/FileMetadata.py:
```diff
@@ -38,56 +40,58 @@ def addMetadataField(self, pname, ptype, credDict):
 
   def deleteMetadataField(self, pname, credDict):
-    """ Remove metadata field
+    """ Remove metadata field (only from user's own VO)
```
please add the method variables to the docscring

zmathe commented on DataManagementSystem/DB/FileCatalogComponents/FileMetadata/FileMetadata.py:
```diff
@@ -38,56 +40,58 @@ def addMetadataField(self, pname, ptype, credDict):
     return result
 
-  def getFileMetadataFields(self, credDict):
+  def getFileMetadataFields(self, credDict, enableStripping=False):
```
same comment as above

zmathe commented on DataManagementSystem/DB/FileCatalogComponents/FileMetadata/FileMetadata.py:
```diff
@@ -100,16 +104,20 @@ def getFileMetadataFields(self, credDict):
   def setMetadata(self, path, metadict, credDict):
-    """ Set the value of a given metadata field for the the given directory path
+    """ Set the value of a given metadata field for the the given directory path.
+        Modified to use fully qualified metadata name.
```
docstring: same as above

zmathe commented on DataManagementSystem/DB/FileCatalogComponents/FileMetadata/FileMetadata.py:
```diff
@@ -143,8 +150,9 @@ def setMetadata(self, path, metadict, credDict):
 
   def removeMetadata(self, path, metadata, credDict):
-    """ Remove the specified metadata for the given file
+    """ Remove the specified metadata for the given file (for user's own VO only)
```
docstring

zmathe commented on DataManagementSystem/DB/FileCatalogComponents/FileMetadata/FileMetadata.py:
```diff
@@ -193,11 +203,11 @@ def __getFileID(self, path):
   def __setFileMetaParameter(self, fileID, metaName, metaValue, credDict):
     """ Set an meta parameter - metadata which is not used in the the data
-        search operations
+        search operations. metaName is VO aware.
```
docstring

zmathe commented on DataManagementSystem/DB/FileCatalogComponents/FileMetadata/FileMetadata.py:
```diff
@@ -238,7 +248,7 @@ def _getFileUserMetadataByID(self, fileIDList, credDict, connection=False):
     return S_OK(metaDict)
 
-  def getFileUserMetadata(self, path, credDict):
+  def getFileUserMetadata(self, path, credDict, enableStripping=False):
```
same as above

atsareg commented:
Returning back to this PR. I agree with Chris and others that the new components should be coded in the way that does not affect the original components. I am sure this can be done.

atsareg commented:
One question: you have put stripVO flag in some methods with False by default. What is the intended usage of it ? The service is always setting to True. I think you can drop it assuming that the suffix is always stripped off. This will also allow you to maintain the same interface for the new components as for the old ones.

martynia commented:
Hi,

I have answered this question on some occasions before. Methods in question are used (not my decision) either by a server RPC call (in which case the suffix is to be stripped off, just before returning to a user) and internally as convenience methods, in which case the suffix is not to be stripped, because convenience methods act on fully qualified names (with the suffix). 

Hope it helps, JM

> On 23 Jan 2020, at 13:22, Andrei Tsaregorodtsev <notifications@github.com> wrote:
> 
> One question: you have put stripVO flag in some methods with False by default. What is the intended usage of it ? The service is always setting to True. I think you can drop it assuming that the suffix is always stripped off. This will also allow you to maintain the same interface for the new components as for the old ones.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/DIRACGrid/DIRAC/pull/4103?email_source=notifications&email_token=AB2KE7LOUQK3EUIRNDTWVLDQ7GKZNA5CNFSM4HSINCOKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJXK6AA#issuecomment-577679104>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AB2KE7K2T7KFWFU2JERLDYTQ7GKZNANCNFSM4HSINCOA>.
>

martynia commented:
@fstagni could you review, please

martynia commented:
@atsareg  

> One question: you have put stripVO flag in some methods with False by default. What is the intended usage of it ? The service is always setting to True. I think you can drop it assuming that the suffix is always stripped off. This will also allow you to maintain the same interface for the new components as for the old ones.

I replied to this. Are you satisfied with my answer or do you have other comments ?

fstagni commented:
IIUC this can be closed, as replaced by more recent PR.
